les/utils/timeutils.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+)
+
+type ThresholdTimer struct {
```

Comment:
Now I understand the purpose of these structures. `ThresholdTimer` is useful but I think it needs a more descriptive name (like `UpdateTimer` because it controls an update process). `DiffTimer` is more weird and I don't think we need a separate structure for that purpose. Maybe the two could be combined: a `ThresholdTimer` (or `UpdateTimer`) with 0 threshold will always call the callback, passing the duration since the last update as a parameter. `DiffTimer.Diff` is also only different because it can operate with an abstime parameter and it does not update `t.Last`. This is my idea:
```
func (t *UpdateTimer) UpdateAt(at mclock.AbsTime, callback func(diff time.Duration) bool) bool {
	t.lock.Lock()
	defer t.lock.Unlock()

	diff := time.Duration(at - t.last)
	if diff < 0 {
		diff = 0
	}
	if diff < t.threshold {
		return false
	}
	if callback(diff) {
		t.last = at
		return true
	}
	return false
}

func (t *UpdateTimer) Update(callback func(diff time.Duration) bool) bool {
	return t.UpdateAt(t.clock.Now(), callback)
}
```
This can do everything that these two structures can (even more because `ThresholdTimer.Update` cannot operate with an external time parameter which might be useful in some other use cases).